django/contrib/auth/middleware.py

```diff
@@ -34,6 +38,89 @@ def process_request(self, request):
         request.auser = partial(auser, request)
 
 
+class LoginRequiredMiddleware(MiddlewareMixin):
+    """
+    Middleware that forces all views to require login by default.
+
+    Views that have login_not_required decorator or LoginNotRequiredMixin mixin
+    will be able to pass through without this validation. Otherwise, it will
+    redirect user to login page or a custom login page if the view has
+    `login_url` or `redirect_field_name` attributes defined.
+    """
+
+    login_url = None
+    permission_denied_message = ""
+    raise_exception = False
+    redirect_field_name = REDIRECT_FIELD_NAME
+
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        if request.user.is_authenticated:
+            return None
+
+        view_class = getattr(view_func, "view_class", None)
+        if view_class and not getattr(view_class, "login_required", True):
+            return None
+
+        if not getattr(view_func, "login_required", True):
+            return None
+
+        return self.handle_no_permission(request, view_func)
+
+    def get_login_url(self, view_func):
+        """
+        Override this method to override the login_url attribute.
+        """
+        login_url = self.login_url or getattr(view_func, "login_url", None)
+        view_class = getattr(view_func, "view_class", None)
+        if view_class:
+            login_url = getattr(view_class, "login_url", None)
+        login_url = login_url or settings.LOGIN_URL
+        if not login_url:
+            raise ImproperlyConfigured(
+                f"{self.__class__.__name__} is missing the login_url attribute. Define "
+                f"{self.__class__.__name__}.login_url, settings.LOGIN_URL, or override "
+                f"{self.__class__.__name__}.get_login_url()."
+            )
+        return str(login_url)
+
+    def get_permission_denied_message(self):
+        """
+        Override this method to override the permission_denied_message attribute.
+        """
+        return self.permission_denied_message
```

Comment:
This is mostly following the [`AccessMixin` api](https://docs.djangoproject.com/en/5.0/topics/auth/default/#django.contrib.auth.mixins.AccessMixin), I have dropped `get_permission_denied_message` and have expanded the docs on the `LoginRequiredMiddleware`.
I can just remove `raise_exception` and `permission_denied_message` and we wait to see if this is requested?

Reply:
Oh, yeah, I think we can drop this feature altogether. It’s just a few lines to override `handle_no_permission()` and add that feature if needed.

Also, it seems somewhat unintentinoal that `LoginRequiredMixin` has the feature. `@login_required` does not, but `@permission_required` does. `LoginRequiredMixin` gets it from `AccessMixin`, perhaps without much thought. At least the inconsistency here points to it not being popular, since otherwise we could expect it to have been added to `@login_required`...

Reply:
Removed 👍